File: algebra_testbed/algebra/physical/physicaloperator.py
#
# Copyright (C) 2011 Akiban Technologies Inc.
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License, version 3,
# as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#

N_STATS = 3
RANDOM_ACCESS = 0
SEQUENTIAL_ACCESS = 1
SORT = 2

class Stats(object):

    def __init__(self):
        self._stats = [0] * N_STATS

    def __getitem__(self, i):
        return self._stats[i]

    def __setitem__(self, i, x):
        self._stats[i] = x

    def merge(self, other):
        merged = Stats()
        for i in xrange(N_STATS):
            merged[i] = self[i] + other[i]
        return merged

class PhysicalOperator(object):

    def __init__(self, input):
        self._stats = Stats()
        self._input = input

    def open(self):
        assert False

    def next(self):
        assert False

    input = property(lambda self: self._input)

    def close(self):
        assert False

    def stats(self):
        assert False

    def count_random_access(self):
        self._stats[RANDOM_ACCESS] += 1

    def count_sequential_access(self):
        self._stats[SEQUENTIAL_ACCESS] += 1

    def count_sort(self, n):
        self._stats[SORT] += n

# Boilerplate for a common pattern: Read from input until a row is
# found, then handle that row.

class SimpleOperator(PhysicalOperator):

    def __init__(self, input):
        PhysicalOperator.__init__(self, input)
        self._input = input

    def open(self):
        self._input.open()

    def next(self):
        output_row = None
        input_row = self._input.next()
        while input_row is not None and output_row is None:
            output_row = self.handle_row(input_row)
            if output_row is None:
                input_row = self._input.next()
        return output_row

    def close(self):
        self._input.close()

    def stats(self):
        return self._stats.merge(self._input.stats())

    def handle_row(self, row):
        assert False

File: src/main/java/com/akiban/qp/physicaloperator/PhysicalOperator.java
/**
 * Copyright (C) 2011 Akiban Technologies Inc.
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 */

package com.akiban.qp.physicaloperator;

import com.akiban.qp.exec.Plannable;
import com.akiban.qp.rowtype.RowType;

import java.util.Collections;
import java.util.List;
import java.util.Set;

public abstract class PhysicalOperator implements Plannable
{
    // I'm not sure I like having this as part of the interface. On one hand, operators like Flatten create new
    // RowTypes and it's handy to get access to those new RowTypes. On the other hand, not all operators do this,
    // and it's conceivable we'll have to invent an operator for which this doesn't make sense, e.g., it creates
    // multiple RowTypes.
    public RowType rowType()
    {
        throw new UnsupportedOperationException();
    }

    @Override
    public List<PhysicalOperator> getInputOperators()
    {
        return Collections.emptyList();
    }

    protected abstract Cursor cursor(StoreAdapter adapter);

    public boolean cursorAbilitiesInclude(CursorAbility ability) {
        return false;
    }

    @Override
    public String describePlan()
    {
        return toString();
    }

    @Override
    public final String describePlan(PhysicalOperator inputOperator)
    {
        StringBuilder buffer = new StringBuilder();
        buffer.append(inputOperator.describePlan());
        buffer.append('\n'); // the newline separator isn't necessarily \u000a, but that's okay
        buffer.append(toString());
        return buffer.toString();
    }
}
